streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java

```diff
@@ -0,0 +1,297 @@
+     * The key for the physical store is the raw key prepended with a fixed-length prefix.
+     */
+    private static class PrefixKeyFormatter {
+        private final byte[] prefix;
+
+        PrefixKeyFormatter(final Bytes prefix) {
+            this.prefix = prefix.get();
+        }
+
+        Bytes forPhysicalStore(final Bytes key) {
+            return key == null ? null : Bytes.wrap(forPhysicalStore(key.get()));
+        }
+
+        byte[] forPhysicalStore(final byte[] key) {
+            final byte[] keyWithPrefix = new byte[prefix.length + key.length];
+            System.arraycopy(prefix, 0, keyWithPrefix, 0, prefix.length);
+            System.arraycopy(key, 0, keyWithPrefix, prefix.length, key.length);
+            return keyWithPrefix;
+        }
+
+        Bytes fromPhysicalStore(final Bytes keyWithPrefix) {
+            return Bytes.wrap(fromPhysicalStore(keyWithPrefix.get()));
+        }
+
+        private byte[] fromPhysicalStore(final byte[] keyWithPrefix) {
+            final int rawKeyLength = keyWithPrefix.length - prefix.length;
+            final byte[] rawKey = new byte[rawKeyLength];
+            System.arraycopy(keyWithPrefix, prefix.length, rawKey, 0, rawKeyLength);
+            return rawKey;
+        }
+
+        Bytes getPrefix() {
+            return Bytes.wrap(prefix);
+        }
+    }
+
+    /**
+     * Converts a {@link KeyValueIterator} which returns keys with prefixes to one which
+     * returns un-prefixed keys.
+     */
+    private static class StrippedPrefixKeyValueIteratorAdapter implements KeyValueIterator<Bytes, byte[]> {
+
+        private final KeyValueIterator<Bytes, byte[]> iteratorWithKeyPrefixes;
+        private final Function<Bytes, Bytes> keyStripper;
+
+        StrippedPrefixKeyValueIteratorAdapter(final KeyValueIterator<Bytes, byte[]> iteratorWithKeyPrefixes,
+                                              final Function<Bytes, Bytes> keyStripper) {
+            this.iteratorWithKeyPrefixes = iteratorWithKeyPrefixes;
+            this.keyStripper = keyStripper;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return iteratorWithKeyPrefixes.hasNext();
+        }
+
+        @Override
+        public KeyValue<Bytes, byte[]> next() {
+            final KeyValue<Bytes, byte[]> next = iteratorWithKeyPrefixes.next();
+            return new KeyValue<>(keyStripper.apply(next.key), next.value);
+        }
+
+        @Override
+        public Bytes peekNextKey() {
+            return keyStripper.apply(iteratorWithKeyPrefixes.peekNextKey());
+        }
+
+        @Override
+        public void remove() {
+            iteratorWithKeyPrefixes.remove();
+        }
+
+        @Override
+        public void close() {
+            iteratorWithKeyPrefixes.close();
+        }
+    }
+
+    private static Bytes serializeLongToBytes(final long l) {
+        return Bytes.wrap(ByteBuffer.allocate(Long.BYTES).putLong(l).array());
```

Comment:
Nit: If we pass this into `PrefixKeyFormatter` and unwrap it there, it seems there is no need to wrap it to begin with?

Reply:
Nice eye. Fixed.